### DISCUSSION TITLE
Some questions about forward dynamic with contact constraint(contact-dynamics.hpp)

### DISCUSSION THREAD

**Discussion by xiaosuyang** [stack-of-tasks/pinocchio#2168] [Q&A] [answered]
Hello. I have some questions after reading contact-dynamics.hpp. 
How to compute a Jacobian of the constraints  take biped robot ，for example?Is that the contact constraint Jocobian and satisfying:
![image](https://github.com/stack-of-tasks/pinocchio/assets/55594017/7ca74018-9430-46dc-a1a2-06c7b8f26bbc)

And what is the gamma(drift of the constraints),is there any information or examples I can refer to?

**Comment by stephane-caron** [comment, marked as answer]
### Constraint Jacobian

You can use the Jacobian of a frame, computed e.g. by [computeFrameJacobian](https://gepettoweb.laas.fr/doc/stack-of-tasks/pinocchio/master/doxygen-html/namespacepinocchio.html#a7b1189aea93a252f4c06d9a7e3c9f5e7), to express your contact constraint $J(q) v = 0$. See also [discussion #1455](https://github.com/stack-of-tasks/pinocchio/issues/1455) to learn more about this topic.

### Drift term

The drift $\gamma = \dot{J}(q, v) v$ that you are looking for will be in `data.a[joint_id]` after calling `forwardKinematics(model,data,q,v,0*v)` (note the zero acceleration `0*v` in the last argument). You can also use `getFrameAcceleration` (or `getFrameClassicalAcceleration`) to get that acceleration in the frame that corresponds to your Jacobian. See [discussion #1395](https://github.com/stack-of-tasks/pinocchio/issues/1395) about this.

(We should also update the documentation to link to this: taking care of this in [#2169](https://github.com/stack-of-tasks/pinocchio/pull/2169).)

You can also look in the code of [TSID](https://github.com/stack-of-tasks/tsid) to learn from another example: [`RobotWrapper::frameClassicAcceleration`](https://github.com/stack-of-tasks/tsid/blob/0a8357568085d43f5a7c518d8cb60958406d0484/src/robots/robot-wrapper.cpp#L295-L305) which is then used e.g. in [`TaskSE3Equality::compute`](https://github.com/stack-of-tasks/tsid/blob/0a8357568085d43f5a7c518d8cb60958406d0484/src/tasks/task-se3-equality.cpp#L155) (unfortunately there is no documentation around those in TSID).

### New example?

There is indeed no example with a $J(q) a + \dot{J}(q, v) v = 0$ constraint at present. If you are currently working on this, feel free to open a PR to propose a minimal one. Other users would thank you for it, and you would get extra :eyes: checking it during the PR review.

**Reply by josyulakrishna** [reply]
Thank you for the answer Stephane. I am very new to using Pinocchio and May I know how to get contact points with a floor and add a floor to the model? I've been searching a lot and I couldn't find any info on this anywhere, some have pointed to adding a floor in gepetto but have not described how to get the contact points. If you could point me to any tutorial or any example code that does basic walking with contact forces I'd be very grateful. Thanks again!

**Reply by stephane-caron** [reply]
Hmm this seems a different question altogether. Can you open a separate [discussion](https://github.com/stack-of-tasks/pinocchio/discussions) so that we keep each discussion thread focused on a single topic?
